src/callbacks/amr_dg2d.jl

```diff
@@ -0,0 +1,330 @@
+  copy!(cache.mortars, mortars)
+
+  # Sanity check
+  if isperiodic(mesh.tree) && nmortars(mortars) == 0
+    @assert ninterfaces(interfaces) == 2 * nelements(dg, cache) ("For 2D and periodic domains and conforming elements, the number of interfaces must be twice the number of elements")
+  end
+
+  return nothing
+end
+
+
+# TODO: Taal compare performance of different implementations
+# Coarsen solution data u for four elements, using L2 projection
+function coarsen_elements!(u::AbstractArray{<:Any,4}, element_id, old_u, old_element_id,
+                           adaptor::LobattoLegendreAdaptorL2, equations, dg)
+  @unpack reverse_upper, reverse_lower = adaptor
+
+  # Store old element ids
+  lower_left_id  = old_element_id
+  lower_right_id = old_element_id + 1
+  upper_left_id  = old_element_id + 2
+  upper_right_id = old_element_id + 3
+
+  @boundscheck begin
+    @assert old_element_id >= 1
+    @assert size(old_u, 1) == nvariables(equations)
+    @assert size(old_u, 2) == nnodes(dg)
+    @assert size(old_u, 3) == nnodes(dg)
+    @assert size(old_u, 4) >= old_element_id + 3
+    @assert     element_id >= 1
+    @assert size(    u, 1) == nvariables(equations)
+    @assert size(    u, 2) == nnodes(dg)
+    @assert size(    u, 3) == nnodes(dg)
+    @assert size(    u, 4) >= element_id
+  end
+
+  for j in eachnode(dg), i in eachnode(dg)
+    acc = zero(get_node_vars(u, equations, dg, i, j, element_id))
+
+    # Project from lower left element
+    for l in eachnode(dg), k in eachnode(dg)
+      acc += get_node_vars(old_u, equations, dg, k, l, lower_left_id) * reverse_lower[i, k] * reverse_lower[j, l]
+    end
+
+    # Project from lower right element
+    for l in eachnode(dg), k in eachnode(dg)
+      acc += get_node_vars(old_u, equations, dg, k, l, lower_right_id) * reverse_upper[i, k] * reverse_lower[j, l]
+    end
+
+    # Project from upper left element
+    for l in eachnode(dg), k in eachnode(dg)
+      acc += get_node_vars(old_u, equations, dg, k, l, upper_left_id) * reverse_lower[i, k] * reverse_upper[j, l]
+    end
+
+    # Project from upper right element
+    for l in eachnode(dg), k in eachnode(dg)
+      acc += get_node_vars(old_u, equations, dg, k, l, upper_right_id) * reverse_upper[i, k] * reverse_upper[j, l]
+    end
+
+    # Update value
+    set_node_vars!(u, acc, equations, dg, i, j, element_id)
+  end
+end
+
+
+function indicator_cache(mesh::TreeMesh{2}, equations, dg::DG, cache)
+
+  indicator_value = Vector{real(dg)}(undef, nelements(dg, cache))
+  return (; indicator_value)
+end
+
+# TODO: Taal refactor, merge the two loops of IndicatorThreeLevel and IndicatorLöhner etc.?
+#       But that would remove the simplest possibility to write that stuff to a file...
+#       We could of course implement some additional logic and workarounds, but is it worth the effort?
+function (indicator::IndicatorThreeLevel)(u::AbstractArray{<:Any,4},
+                                          mesh::TreeMesh{2}, equations, dg::DG, cache)
```

Comment:
We could use different names for `cache` if different types of caches are used somewhere. I strongly disagree with using type annotations, since that will not be reliable using the current infrastructure. For simplicity, I'm using `NamedTuple`s for every kind of cache. The exact type of these depends on the ingredients used, e.g. the volume integral etc. for DG methods. Hence, uusing dispatch here seems to be too restrictive and not really helpful for me. 

Nevertheless, we can improve the readability of the code by choosing slightly more descriptive names and adding more comments.

Reply:
OK. Using more descriptive names would probably suffice, since then the cache origin is already clear